### ISSUE TITLE
[OJS] Upgrading to OJS3: multiple issues with ongoing reviews

### ISSUE THREAD

**Issue by ajnyga** [pkp/pkp-lib#2173]
Described here: http://forum.pkp.sfu.ca/t/ojs3-upgrade-multiple-issues-with-ongoing-reviews/27563

OJS 3.0.1

**Comment by ajnyga** [comment]
Our site upgraded from OJS2.4.8.1 to OJS3.0.1.

There are at least two issues our journals are seeing with the ongoing reviews.

    All the "Files To Be Reviewed" are unselected in all ongoing reviews. The file is there if I edit an active assignment, but it is not selected to be viewed to the reviewer.

    All the "Reviewer files" that the reviewers have sent are missing from the grid. (probably a similar issue as with the missing revisions)

    All cancelled reviews are listed with "Review submitted" stage.
    This happens probably because in OJS3 cancelled reviews are deleted from the review_assignments table. However, in OJS2 these only get field cancelled set to 1. Probably the upgrade script should delete these from the table? Are all review_assignments with cancelled set to 1 safe to delete?

    All reviews that are active are marked as Overdue
    This is because in review_assignments the date_response_due is not set at all for the reviews initiated in OJS2 so OJS3 is thinking that all of these responses are late

Any ideas how to fix this automatically for 45 journals and circa 200 ongoing reviews?

**Comment by ajnyga** [comment]
Concerning item 2:
The reviewer files are in submission_files table, but there are several problems
file_stage is 4, should be 5
assoc_id is missing, should be review_id from review_assignments
assoc_type is missing, should be 517

**Comment by ajnyga** [comment]
For item 4 the best solution I can think of is to set date_response_due to match the value of date_due.

**Comment by bozana** [comment]
OK, lets take a look at the issues here.. :-)

**Comment by bozana** [comment]
@ajnyga, regarding the issue no. 1: Do you then see the error message "No Files Selected. You have not selected any files for the reviewer to review."? Is the `file_id` in the DB table `review_assignments` for that review assignment the same as the listed file ID?

**Comment by ajnyga** [comment]
Hi,

Yes, I see that error if I edit an assignment. This is not a critical issue, because the editor can fix this easily just by selecting the file, but maybe something that could be taken into account in the migration script.

**Comment by ajnyga** [comment]
Regarding issue 2, I fixed this with sql and a script I used to move and rename the files. I think the central problem there is the wrong file stage.

**Comment by bozana** [comment]
I think that there should be one entry for each review assignment from OJS 2.4.x in the new table `review_files`. -- There was always a review file in OJS 2.4.x, in table `articles`, column `review_file_id`. The stage_id for that file with id = review_file_id was migrated, but the entry in review_files table was not made/considered.

**Comment by ajnyga** [comment]
Issue 1: the file itself is available when I go to edit a review assignment, it just not selected. 
So it seems that the selection of whether the editor want's to show the file to the reviewer is not migrated correctly. I can compare the preupdate OJS2 tables to the updated OJS3 ones.

**Comment by bozana** [comment]
This checkbox (file selection) looks if there is a file in the DB table `review_files`. In OJS 2.4.x there was no such mechanism -- there was always a file for the reviewer, saved in the articles table review_file_id column. This column should be migrated to the review_files table, I believe.

**Comment by bozana** [comment]
@ajnyga, concerning the problem 2 above: is is so that only those migrated reviewer files are in the table review_assignments, column reviewer_file_id? -- just to double check, that I can repair it looking at this column...

**Comment by ajnyga** [comment]
I have to see if I can find out somehow, because I already fixed this post upgrade with my own script...

**Comment by ajnyga** [comment]
Ok, so sorry @bozana can you rephrase your question?

**Comment by bozana** [comment]
Do you have something in your column `review_file_id` in `review_assignments` table? If so, are those the migrated (from OJS 2.4.x) reviewer files ids?

**Comment by ajnyga** [comment]
Yes I have and the values there seem to match the values of our old OJS2 database.

**Comment by bozana** [comment]
Thanks!!!

**Comment by ajnyga** [comment]
And the id there is at least in the three first cases the same as in the submission_files table.

**Comment by bozana** [comment]
PRs:
1.
2.
ojs master: https://github.com/pkp/ojs/pull/1236

**Comment by bozana** [comment]
@asmecher, could you take a look at the PR above, that fixes the problem number 2 in this issue? At the end I remove the reviewer_file_ids from review_assignments, as it is in OJS 3. If they should stay -- to be on the safe side -- I will have to consider that so that the SQLs are not executed several times, every time when doing upgrade... Thanks!

**Comment by ajnyga** [comment]
Thanks, I will run  <query>UPDATE review_assignments SET reviewer_file_id = NULL WHERE reviewer_file_id IS NOT NULL</query> in our database to prepare for 3.1 upgrade. But do we need the whole `reviewer_file_id` field after this?

**Comment by bozana** [comment]
@ajnyga, please wait till Alec answers. I can then let you know... Do not change anything in you solution till this is merged... :-)

**Comment by asmecher** [comment]
I think that field should be removed; I don't think it's relevant in OJS3.

**Comment by bozana** [comment]
@asmecher, could you take a look at this PR that solves number 1) in this issue: pkp/ojs#1282 ?

**Comment by bozana** [comment]
@asmecher, would it be ok to set date_response_due in review_assignments to date_due if the first one is NULL i.e. migrated from OJS 2.4.x? (s. issue 4. above).

**Comment by asmecher** [comment]
Yes, I think that's a good solution.

**Comment by ajnyga** [comment]
At least I have not heard about any problems related to this solution.

**Comment by ajnyga** [comment]
Hi @bozana 

Regarding issue 2 here. I found out an interesting problem with the assumption that there is only one reviewer file present. 

It turned out that some reviewers have uploaded several attached files in OJS2. Because the UI there was not very clear about the file revisions they have not noticed that they are basically loading a new version of a same file (file_id).

So after upgrading to OJS3, only the last of the files is showing because OJS3 is treating it as the final revision of the reviewer file.

It is the exact same problem like the one with the author revisions described here:
https://github.com/pkp/pkp-lib/issues/2097#issuecomment-268255269

Here is one of the cases from our old OJS2 installation:
![screenshot_6](https://cloud.githubusercontent.com/assets/16347527/23802502/423460c4-05bc-11e7-89dd-6a177bee09a5.png)
Those are two totally different files, but there is not much in the UI suggesting that OJS2 is thinking they are two separate files. Also, I do think that by selecting the show option from both, the author would also see them both in OJS2?

So, making the assumption in the OJS3 upgrade that there is only one file (possibly with revisions) attached to a review is probably going to lead to some problems. In our case I just have to deal with these cases, but maybe this could be considered somehow for the users doing their upgrade later?

Also, does the upgrade script consider the revisions of those files? Because the script I did to solve issue 2 after our upgrade did not, I just went with the original assumption of having a single file.

**Comment by bozana** [comment]
Hi @ajnyga 
Thanks a lot! Yes, we realized that too, when doing an upgrade :-\ Thus, I fixed that, s. the second pair of PRs above under number 2. i.e. https://github.com/pkp/ojs/pull/1292

**Comment by ajnyga** [comment]
Ah yes, I see now. Thank you, good to know. Luckily we did not have too much of these cases. But I do have to figure out in May how these affect our installation during an upgrade. Probably no issue there.

The second point is of course due to wrong use of OJS2, but I do think that it is not that uncommon (to submit separate files and not realizing the system is creating revisions).

**Comment by bozana** [comment]
@ajnyga, I think the second point we cannot fix... right?

**Comment by ajnyga** [comment]
Hi, definitely not because there could be cases where the reviewer has used OJS2 the right way and has sent an actual revision of the same file. 

However, in OJS3, would the editor be able to see the different revisions of the reviewer file?

**Comment by bozana** [comment]
Hmmm... I would say as all the others file revisions... :-)

**Comment by bozana** [comment]
Here only the number 3 is left:

> All cancelled reviews are listed with "Review submitted" stage.
> This happens probably because in OJS3 cancelled reviews are deleted from the review_assignments table. However, in OJS2 these only get field cancelled set to 1. Probably the upgrade script should delete these from the table? Are all review_assignments with cancelled set to 1 safe to delete?

@asmecher, would you also agree to delete the canceled reviews when migrating?

**Comment by bozana** [comment]
@asmecher, the last PR for this issue, solving number 3 above: https://github.com/pkp/ojs/pull/1317. Could you please take a look?

**Comment by bozana** [comment]
@asmecher, here are the new PRs, that also remove the review assignments column "cancelled", this time against master branch, because it is not just a fix:
- ojs: https://github.com/pkp/ojs/pull/1321
- reviewReport: https://github.com/pkp/reviewReport/pull/2
- pkp-lib: https://github.com/pkp/pkp-lib/pull/2366
S. also comments here: https://github.com/pkp/ojs/pull/1317#issuecomment-287436727

**Comment by asmecher** [comment]
Looks good -- thanks, @bozana! Please go ahead. Agreed that this should go into master, not `ojs-stable-3_0_2`, and some of the filenames etc. need updating to reflect that.

**Comment by bozana** [comment]
Just the submodule plugins/report/reviewReport to be merged, thus closing...
